### PR TITLE
constellation: Simplify the scheduler so it doesn't create a thread for each event

### LINKED ISSUE #11268: Avoid creating a thread every single event we receive
We're doing this right now, for some reason I can't understand. We spawn a single thread just to sleep, ensure we're not cancelled, and send an empty message.

If we're going to use threads, we should at least use a thread pool, but probably there's a smarter way to do this, like a single thread that sleeps just the necessary time to dispatch the next event, or something.

I plan to take a look at this, but my time is actually constrained a lot, so if someone wants to investigate and refactor this, please do!

### PR DESCRIPTION
**Pull request by emilio**
Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data:
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy --faster` does not report any errors
- [x] These changes fix #11268 

Either:
- [x] These changes do not require tests because it's a refactoring

Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. 

---

This PR builds on top of #11272 (it couldn't be reopened because I pushed to the branch before doing it).

r? @asajeffrey 

cc @metajack @pcwalton 

Concretely:

@metajack: This uses an auxiliary thread to take care of timeouts. It's not optimal (optimal would be https://github.com/rust-lang/rfcs/issues/962), but it's way better than anything we had, and the other option (passing the thread handle around and `unpark()` the thread when appropiate) wouldn't work across IPC.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11279)

### CODE REVIEW

**Comment by asajeffrey** [comment]
My **completely untested** implementation, for comparison: https://github.com/asajeffrey/servo/tree/scheduler-with-fewer-threads

**Comment by asajeffrey** [comment]
Some IRC discussion: http://logs.glob.uno/?c=mozilla%23servo&s=19+May+2016&e=19+May+2016#c433870

TL;DR (but really it's very short): I'm concerned that this code may deadlock, or at least block for an unbounded amount of time, due to a race condition. Fixing it (as far as we can see) requires spinning on an atomic, since there's no test-and-park primitive.

**Comment by asajeffrey** [comment]
My now-tested implementation: #11283.

**Comment by emilio** [comment]
To be clear, I'm fine landing any of both approaches. I think this one would be a bit easier to migrate if https://github.com/rust-lang/rfcs/issues/962 is implemented, but that's not a huge concern.

On the other side, after reading the docs more carefully, I think the discussion we had on IRC was wrong. This can't make us sleep too much. The point @asajeffrey was making was that if we called `timeout_thread.unpark()` before the thread started to park, it will sleep the timeout before returning.

From the docs:

```
park: Blocks unless or until the current thread's token is made available. 
```

(notice the "unless" bit).

That would mean that the worst case for this is the unlikely case that we call `unpark()` when the timeout thread's `unpark_timeout` has already ended, and thus **the next time** it would return automatically (since the token is available).

This is not a problem for this implementation, since once the timeout is raised, we'll re-check times before checking that again, and block, so this becomes a no-op.

In fact, I think this can't even be the case, since `mpsc::recv()` uses `park()` checking an atomic flag internally.

**Comment by asajeffrey** [comment]
Ah, this makes life a lot easier, I'll try again with this semantics of park/unpark.

**Comment by asajeffrey** [comment]
We decided to go for #11283.
